rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,146 @@
+
+## Benefits
+
+None of the registered types will expose their encrypted details. The saved
+objects with their unencrypted attributes could still be obtained and searched
+on. The wrapper will follow all the security and spaces filtering of saved
+objects so that only users with appropriate permissions will be able to obtain
+the scrubbed objects or _save_ objects with encrypted attributes.
+
+## Possible Problems
+
+Consumers of scrubbed objects may try to modify them causing the encrypted
+attributes to be dropped. But this could be prevented by disallowing updates
+from the default saved objects client. And instead force all updates to go
+through the plugin directly.
+
+## Decrypting attributes
+
+In order to get the resulting decrypted attributes back the plugin would require
+a method be registered which would be passed the saved object including the
+decrypted attributes then return an invokable method that would perform the
+retrieval of the saved object the decryption and invoking of the desired method.
+
+```JS
+const actionMethod = server.plugins.encryptedAttributes.buildDecryptionContext(id, type, (savedObject) => {
+  const { username, password } = savedObject.secretsParams;
+  ... // do integration with third party
+});
+... // later
+actionMethod(); // retrieve, decrypt, and invoke above method
+```
+
+## Benefits
+
+No explicit access to a method that takes in an encrypted string exists. And no
+access to the private key used is exposed through the plugin. If the type was not
+registered no decryption is possible. No need to handle the saved object with
+the encrypted attributes reducing the risk of accidentally returning it in a
+handler.
+
+# Drawbacks
+
+- This would add the ability to encrypt data on any saved object in the system.
+- Possibly add performance issues to retrieving saved objects.
+- Might be complicated to test along with spaces and security
+- Could be too complicated in general
+- The attributes that are encrypted have to be defined and if they change they
+  need to be migrated.
+
+# Alternatives
+
+Only allow this to be used within the Actions service itself where the details
+of the saved object are handled there directly. And the saved objects are
+`hidden` but still use the security and spaces wrappers.
+
+# Adoption strategy
+
+Integration should be pretty easy which would include depending on the plugin.
+And registering the desired saved object type with it. And building and using
+the desired decryption contexts.
+
+# How we teach this
+
+`encryptedAttributes` as the name of the `thing` where it's seen as a separate
+extension on top of the saved object service.
+
+Provide a readme.md in the plugin directory with examples for how to depend on
+the plugin and define a type has hidden attributes.
+
+# Unresolved questions
+
+- Is `encryptedAttributes` an acceptable name for the plugin
+- Is building a method that receives the object with the decrypted attributes
```

Comment:
Yes, that is what I am referring to, and I am a little concerned that doing it this way might lead to hacks to work around the syntax or coarse the syntax for other purposes or other code styles perhaps with promises or something. I think this syntax should be usable in most circumstances, just wanted another opinion.

Reply:
I think `Promise` and `async`/`await` are becoming the new "standard" so using a callback here does seem a little out of place. If our intent is to try to use a callback here to limit the scope of the decrypted values, it'd be rather easy for a consumer to turn this into a Promise and negate that potential benefit.

Reply:
It makes sense to turn this into a Promise, but I'm thinking more of returning a function that returns a promise so that saved object is not immediately decrypted but only decrypted when it is closer to use. The point of the code is to try and capture the audit log closer to the usage of the saved object not sure that is entirely possible.

Reply:
On second thought it makes no difference, I should just return a promise to the object then the code would become.

```JS
const fullObject = await server.plugins.encryptedAttributes.get(it, type);
const { username, password } = fullObject.secretsParams;
```